# Business Understanding
## Overview of the business
A company that is seeking to expand into the aviation industry and wants buy aircrafts for its new division

## Business Problem

Being new in the aviation industry the company is unaware of the potential risks that are associated with aircrafts 

## Business Objectives
The company wants to understand the potential risks it may face in the aviation industry
The company wants to obtain a list of recommended aircrafts that are considered to have low risks in regards to its operational use

## Success Criteria
To obtain a list of aircrafts that are considered to have low risks

To outline the potential risks the company may experience in the aviation industry

To produce relevant visualisations that provide insights on aircraft and the aviation industry

## Project plan

To go through a dataset that contains accidents that have occurred in the aviation industry and to obtain relevant insights that can be able to solve the business problem






# Familization with the dataset

In [50]:
# Importing the pandas and matplotlib libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import re
from geopy.geocoders import Nominatim
%matplotlib inline

In [51]:
pd.options.display.max_columns = None

Loading the dataset to be used


In [52]:
aviation_data_df = pd.read_csv("data/AviationData.csv", encoding='latin-1', dtype='str')


In [53]:
usstate_codes_df = pd.read_csv("data/USState_Codes.csv")

Checking the head, tail and a sample of the dataset so to get an understanding of the dataset we are working with


In [54]:
aviation_data_df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name Injury.Severity Aircraft.damage Aircraft.Category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  Registration.Number      Make     Model Amateur.Built Number.of.Engines  \
0              NC6404   Stinson     108-3            No                 1   
1              N5069P     Piper  PA24-180            No                 1   
2              N5142R    Cessna      172M            No                 1   
3              N1168J  Rockwell       112            No                 1   
4               N15NY    Cessna       501            No               NaN   

     Engine.Type FAR.Description Schedule Purpose.of.flight Air.carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

  Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
0                    2                      0                    0   
1                    4                      0                    0   
2                    3                    NaN                  NaN   
3                    2                      0                    0   
4                    1                      2                  NaN   

  Total.Uninjured Weather.Condition Broad.phase.of.flight   Report.Status  \
0               0               UNK                Cruise  Probable Cause   
1               0               UNK               Unknown  Probable Cause   
2             NaN               IMC                Cruise  Probable Cause   
3               0               IMC                Cruise  Probable Cause   
4               0               VMC              Approach  Probable Cause   

  Publication.Date  
0              NaN  
1       19-09-1996  
2       26-02-2007  
3       12-09-2000  
4       16-04-1980

In [55]:
aviation_data_df.tail()

Event.Id Investigation.Type Accident.Number  Event.Date  \
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

            Location        Country Latitude Longitude Airport.Code  \
88884  Annapolis, MD  United States      NaN       NaN          NaN   
88885    Hampton, NH  United States      NaN       NaN          NaN   
88886     Payson, AZ  United States  341525N  1112021W          PAN   
88887     Morgan, UT  United States      NaN       NaN          NaN   
88888     Athens, GA  United States      NaN       NaN          NaN   

      Airport.Name Injury.Severity Aircraft.damage Aircraft.Category  \
88884          NaN           Minor             NaN               NaN   
88885          NaN             NaN             NaN               NaN   
88886       PAYSON       Non-Fatal     Substantial          Airplane   
88887          NaN             NaN             NaN               NaN   
88888          NaN           Minor             NaN               NaN   

      Registration.Number                        Make      Model  \
88884              N1867H                       PIPER  PA-28-151   
88885              N2895Z                    BELLANCA       7ECA   
88886              N749PJ  AMERICAN CHAMPION AIRCRAFT      8GCBC   
88887              N210CU                      CESSNA       210N   
88888              N9026P                       PIPER  PA-24-260   

      Amateur.Built Number.of.Engines Engine.Type FAR.Description Schedule  \
88884            No               NaN         NaN             091      NaN   
88885            No               NaN         NaN             NaN      NaN   
88886            No                 1         NaN             091      NaN   
88887            No               NaN         NaN             091      NaN   
88888            No               NaN         NaN             091      NaN   

      Purpose.of.flight         Air.carrier Total.Fatal.Injuries  \
88884          Personal                 NaN                    0   
88885               NaN                 NaN                    0   
88886          Personal                 NaN                    0   
88887          Personal  MC CESSNA 210N LLC                    0   
88888          Personal                 NaN                    0   

      Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
88884                      1                    0               0   
88885                      0                    0               0   
88886                      0                    0               1   
88887                      0                    0               0   
88888                      1                    0               1   

      Weather.Condition Broad.phase.of.flight Report.Status Publication.Date  
88884               NaN                   NaN           NaN       29-12-2022  
88885               NaN                   NaN           NaN              NaN  
88886               VMC                   NaN           NaN       27-12-2022  
88887               NaN                   NaN           NaN              NaN  
88888               NaN                   NaN           NaN       30-12-2022

In [56]:
aviation_data_df.sample(10)

Event.Id Investigation.Type Accident.Number  Event.Date  \
74939  20140414X14114           Accident      WPR14FA167  2014-04-13   
6153   20001214X44389           Accident      ANC83LA162  1983-09-04   
25063  20001212X22774           Accident      MIA90LA093  1990-03-14   
24263  20001213X29551           Accident      CHI90DEM01  1989-10-13   
48156  20001212X20832           Accident      IAD00TA040  2000-04-11   
52206  20020308X00321           Accident      ATL02FA058  2002-03-03   
82425  20181009X31826           Accident      GAA19CA010  2018-10-08   
57711  20041029X01725           Accident      ATL05FA008  2004-10-23   
43742  20001211X09892           Accident      LAX98LA139  1998-04-20   
73503  20130508X01807           Accident      ERA13CA231  2013-05-04   

               Location        Country   Latitude   Longitude Airport.Code  \
74939          Reno, NV  United States    039405N    0119523W          RTS   
6153       SKWENTNA, AK  United States        NaN         NaN          SKW   
25063          MIMS, FL  United States        NaN         NaN          NaN   
24263      GRAYLING, MI  United States        NaN         NaN          55D   
48156     CLEVELAND, OH  United States        NaN         NaN          CLE   
52206         Mexia, TX  United States  31.619167  -96.502778          LXY   
82425       Crowley, LA  United States    030318N    0921445W          PVT   
57711     FLAT ROCK, NC  United States  35.315556  -82.353056          0A7   
43742  GRAND CANYON, AZ  United States        NaN         NaN         KGCN   
73503       Arcadia, NY  United States    043739N    0077247W         NONE   

               Airport.Name Injury.Severity Aircraft.damage Aircraft.Category  \
74939            RENO/STEAD           Fatal     Substantial        Helicopter   
6153                SKWETNA       Non-Fatal     Substantial               NaN   
25063                   NaN       Non-Fatal     Substantial               NaN   
24263          GRAYLING AAF        Fatal(1)       Destroyed               NaN   
48156     CLEVELAND-HOPKINS       Non-Fatal     Substantial               NaN   
52206      Limestone County        Fatal(1)       Destroyed               NaN   
82425               Private       Non-Fatal     Substantial          Airplane   
57711        HENDERSONVILLE        Fatal(1)     Substantial               NaN   
43742  GRAND CANYON AIRPORT       Non-Fatal     Substantial               NaN   
73503         Private Strip       Non-Fatal     Substantial          Airplane   

      Registration.Number         Make      Model Amateur.Built  \
74939               N78CS        SANDS  HELICYCLE           Yes   
6153               N4097Z        Piper  PA-18-150            No   
25063              N66949       Cessna        152            No   
24263              N3905U       Cessna       150E            No   
48156              N6183T       Cessna      182RG            No   
52206              N7272D        Beech      BE-60            No   
82425              N1500E  Air Tractor     AT 402            No   
57711              N18303        Beech       F33A            No   
43742              N8533F         Bell       206B            No   
73503              N2143R       CESSNA       182G            No   

      Number.of.Engines    Engine.Type FAR.Description Schedule  \
74939                 1    Turbo Shaft             091      NaN   
6153                  1  Reciprocating             NaN      NaN   
25063                 1  Reciprocating             NaN      NaN   
24263                 1  Reciprocating             NaN      NaN   
48156                 1  Reciprocating             NaN      NaN   
52206                 2  Reciprocating             NaN      NaN   
82425                 1     Turbo Prop             137      NaN   
57711                 1  Reciprocating             NaN      NaN   
43742                 1    Turbo Shaft             NaN      NaN   
73503                 1  Reciprocating             091      N

In [57]:
aviation_data_df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [58]:
aviation_data_df.shape

(88889, 31)

In [59]:
len(aviation_data_df.columns)

31

In [60]:
usstate_codes_df.head()

US_State Abbreviation
0     Alabama           AL
1      Alaska           AK
2     Arizona           AZ
3    Arkansas           AR
4  California           CA

In [61]:
usstate_codes_df.tail()

US_State Abbreviation
57  Virgin Islands           VI
58   Washington_DC           DC
59  Gulf of mexico           GM
60  Atlantic ocean           AO
61   Pacific ocean           PO

In [62]:
usstate_codes_df.shape

(62, 2)

In [63]:
usstate_codes_df.columns

Index(['US_State', 'Abbreviation'], dtype='object')

# DATA UNDERSTANDING



The data folder contains two data sets

## Aviation dataset
This dataset contains data of reported aviation accidents and information regarding the accident reported.

The dataset contains 31 columns with each column containing information on the accident reported.

Some of the columns in the dataset include:date of the accident, the severity of the accident, country where the accident occurred, aircraft damage, aircraft type and model etc

The dataset contains 88889 records

## US State Codes dataset
This dataset contains information on the full name of abbreviated us state codes

The dataset contains two columns i.e the US state and its abbrevaiation

The dataset contains 62 records





# Data cleaning

## Aviation Dataset

In [64]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Event.Id                88889 non-null  object
 1   Investigation.Type      88889 non-null  object
 2   Accident.Number         88889 non-null  object
 3   Event.Date              88889 non-null  object
 4   Location                88837 non-null  object
 5   Country                 88663 non-null  object
 6   Latitude                34382 non-null  object
 7   Longitude               34373 non-null  object
 8   Airport.Code            50249 non-null  object
 9   Airport.Name            52790 non-null  object
 10  Injury.Severity         87889 non-null  object
 11  Aircraft.damage         85695 non-null  object
 12  Aircraft.Category       32287 non-null  object
 13  Registration.Number     87572 non-null  object
 14  Make                    88826 non-null  object
 15  Mo

Using pd.info we can observe the following

a) every column contain string values

b)They are columns with null values

c)Columns names contain different structures and need to be standardized

### Converting column names to a simple standardized form

a)column names to be converted to lower case

b)column names to be seperated by underscore


In [65]:
#converting the columns into lower case
aviation_data_df.columns=aviation_data_df.columns.str.lower()

In [66]:
#seperating column names with an underscore rather than with a period

aviation_data_df.columns=aviation_data_df.columns.str.replace('.','_')

aviation_data_df.columns

Index(['event_id', 'investigation_type', 'accident_number', 'event_date',
       'location', 'country', 'latitude', 'longitude', 'airport_code',
       'airport_name', 'injury_severity', 'aircraft_damage',
       'aircraft_category', 'registration_number', 'make', 'model',
       'amateur_built', 'number_of_engines', 'engine_type', 'far_description',
       'schedule', 'purpose_of_flight', 'air_carrier', 'total_fatal_injuries',
       'total_serious_injuries', 'total_minor_injuries', 'total_uninjured',
       'weather_condition', 'broad_phase_of_flight', 'report_status',
       'publication_date'],
      dtype='object')

In [67]:
aviation_data_df.head()

event_id investigation_type accident_number  event_date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          location        country   latitude   longitude airport_code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built number_of_engines  \
0              NC6404   Stinson     108-3            No                 1   
1              N5069P     Piper  PA24-180            No                 1   
2              N5142R    Cessna      172M            No                 1   
3              N1168J  Rockwell       112            No                 1   
4               N15NY    Cessna       501            No               NaN   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

  total_fatal_injuries total_serious_injuries total_minor_injuries  \
0                    2                      0                    0   
1                    4                      0                    0   
2                    3                    NaN                  NaN   
3                    2                      0                    0   
4                    1                      2                  NaN   

  total_uninjured weather_condition broad_phase_of_flight   report_status  \
0               0               UNK                Cruise  Probable Cause   
1               0               UNK               Unknown  Probable Cause   
2             NaN               IMC                Cruise  Probable Cause   
3               0               IMC                Cruise  Probable Cause   
4               0               VMC              Approach  Probable Cause   

  publication_date  
0              NaN  
1       19-09-1996  
2       26-02-2007  
3       12-09-2000  
4       16-04-1980

### Converting columns that contain string values to respective dtypes

This columns were all converted to string values while reading the csv so to clear error generated.Therefore we need to go through each column and with intuition convert some of the columns back to their original data type

In [68]:
aviation_data_df.sample(10)

event_id investigation_type accident_number  event_date  \
11806  20001214X36483           Accident      FTW85LA241  1985-05-31   
67690  20100204X74536           Incident      ENG10RA014  2010-02-03   
10056  20001214X41275           Accident      ATL85LA012  1984-10-14   
55214  20030808X01291           Accident      NYC03LA175  2003-07-31   
819    20020917X01858           Accident      CHI82DA119  1982-04-11   
77188  20150903X93747           Accident      GAA15CA249  2015-08-20   
20819  20001213X26262           Accident      LAX88FA252  1988-07-12   
42539  20001208X08732           Accident      NYC97LA169  1997-08-23   
64279  20080415X00482           Accident      DEN08LA071  2008-04-04   
83589  20190904X11028           Accident      ANC19LA049  2019-07-17   

                          location        country latitude   longitude  \
11806               FORT WORTH, TX  United States      NaN         NaN   
67690  Bangkok, Thailand, Thailand       Thailand      NaN         NaN   
10056                    WIRTZ, VA  United States      NaN         NaN   
55214               WHITE POST, VA  United States  39.0625  -78.091389   
819                       NOVA, OH  United States      NaN         NaN   
77188                   Tucson, AZ  United States  032649N    1105612W   
20819                   LOMPOC, CA  United States      NaN         NaN   
42539              SPENCERPORT, NY  United States      NaN         NaN   
64279                 La Porte, TX  United States  294017N    0095425W   
83589                Talkeetna, AK  United States  062582N    1504456W   

      airport_code                airport_name injury_severity  \
11806         17TA                    HELIDYNE       Non-Fatal   
67690          BKK  Suvarnabhumi International             NaN   
10056          NaN                BOWMAN FIELD       Non-Fatal   
55214         3VA7          White Post Airport       Non-Fatal   
819            NaN          PRIVATE FARM STRIP       Non-Fatal   
77188          TUS                 TUCSON INTL       Non-Fatal   
20819          LPC                      LOMPOC        Fatal(1)   
42539         NONE                 SPENCERPORT       Non-Fatal   
64279          T41          La Porte Municipal       Non-Fatal   
83589          NaN                         NaN       Non-Fatal   

      aircraft_damage aircraft_category registration_number     make  \
11806     Substantial               NaN              N1072Y     Bell   
67690           Minor          Airplane              HS-TGB   BOEING   
10056     Substantial               NaN              N6348G   Cessna   
55214     Substantial               NaN              N97168  Stinson   
819       Substantial          Airplane              N3402Z    Piper   
77188     Substantial          Airplane              N944LB  STINSON   
20819       Destroyed               NaN               N27EB    Beech   
42539     Substantial               NaN              N7682H    Piper   
64279     Substantial        Helicopter              N40289   Hiller   
83589     Substantial          Airplane              N121EE   Cessna   

           model amateur_built number_of_engines    engine_type  \
11806      206-B            No                 1    Turbo Shaft   
67690        747            No                 4            NaN   
10056        150            No                 1  Reciprocating   
55214        108            No                 1  Reciprocating   
819    PA-22-160            No                 1  Reciprocating   
77188       L 5E            No                 1  Reciprocating   
20819        A35            No                 1  Reciprocating   
42539      PA-12            No                 1  Reciprocating   
64279     UH-12E            No                 1  Reciprocating   
83589       A185            No               NaN  Reciprocating   

                 far_description schedule purpose_of_flight      air_carrier  \
11806                        NaN      NaN           Unknown        

#### number_of_engines,total_fatal_injuries,total_minor_injuries,total_serious_injuries and total_uninjured to interger values
Taking a sample of the dataset we can safely deduce that the following columns originally contained interger values

number_of_engines,total_fatal_injuries,total_minor_injuries,total_serious_injuries and total_uninjured

In [69]:
#We first convert the columns above to numeric
aviation_data_df['number_of_engines'] = pd.to_numeric(aviation_data_df['number_of_engines'],errors= 'coerce')
aviation_data_df['total_fatal_injuries'] = pd.to_numeric(aviation_data_df['total_fatal_injuries'], errors = 'coerce')
aviation_data_df['total_serious_injuries'] = pd.to_numeric(aviation_data_df['total_serious_injuries'], errors = 'coerce')
aviation_data_df['total_minor_injuries'] = pd.to_numeric(aviation_data_df['total_minor_injuries'], errors = 'coerce')
aviation_data_df['total_uninjured'] = pd.to_numeric(aviation_data_df['total_uninjured'], errors = 'coerce')


In [70]:

aviation_data_df[['number_of_engines','total_fatal_injuries','total_serious_injuries','total_minor_injuries','total_uninjured']]=aviation_data_df[['number_of_engines','total_fatal_injuries','total_serious_injuries','total_minor_injuries','total_uninjured']].astype('Int32')

In [71]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   event_id                88889 non-null  object
 1   investigation_type      88889 non-null  object
 2   accident_number         88889 non-null  object
 3   event_date              88889 non-null  object
 4   location                88837 non-null  object
 5   country                 88663 non-null  object
 6   latitude                34382 non-null  object
 7   longitude               34373 non-null  object
 8   airport_code            50249 non-null  object
 9   airport_name            52790 non-null  object
 10  injury_severity         87889 non-null  object
 11  aircraft_damage         85695 non-null  object
 12  aircraft_category       32287 non-null  object
 13  registration_number     87572 non-null  object
 14  make                    88826 non-null  object
 15  mo

In [72]:
aviation_data_df.head()

event_id investigation_type accident_number  event_date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          location        country   latitude   longitude airport_code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built  number_of_engines  \
0              NC6404   Stinson     108-3            No                  1   
1              N5069P     Piper  PA24-180            No                  1   
2              N5142R    Cessna      172M            No                  1   
3              N1168J  Rockwell       112            No                  1   
4               N15NY    Cessna       501            No               <NA>   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

   total_fatal_injuries  total_serious_injuries  total_minor_injuries  \
0                     2                       0                     0   
1                     4                       0                     0   
2                     3                    <NA>                  <NA>   
3                     2                       0                     0   
4                     1                       2                  <NA>   

   total_uninjured weather_condition broad_phase_of_flight   report_status  \
0                0               UNK                Cruise  Probable Cause   
1                0               UNK               Unknown  Probable Cause   
2             <NA>               IMC                Cruise  Probable Cause   
3                0               IMC                Cruise  Probable Cause   
4                0               VMC              Approach  Probable Cause   

  publication_date  
0              NaN  
1       19-09-1996  
2       26-02-2007  
3       12-09-2000  
4       16-04-1980

#### converting event_date and publication_date to pandas.Datetime

In [73]:
#Converting columns that contain dates into pandas.datetime
# that is event_date and publication_date


aviation_data_df['event_date']=pd.to_datetime(aviation_data_df['event_date'],format="%Y-%m-%d" ,yearfirst= True, errors= 'coerce')



In [74]:
aviation_data_df['publication_date']=pd.to_datetime(aviation_data_df['publication_date'],format="%d-%m-%Y", yearfirst= True, errors='coerce')

In [75]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   event_id                88889 non-null  object        
 1   investigation_type      88889 non-null  object        
 2   accident_number         88889 non-null  object        
 3   event_date              88889 non-null  datetime64[ns]
 4   location                88837 non-null  object        
 5   country                 88663 non-null  object        
 6   latitude                34382 non-null  object        
 7   longitude               34373 non-null  object        
 8   airport_code            50249 non-null  object        
 9   airport_name            52790 non-null  object        
 10  injury_severity         87889 non-null  object        
 11  aircraft_damage         85695 non-null  object        
 12  aircraft_category       32287 non-null  object

In [76]:
aviation_data_df[['event_date','publication_date']].sample(20)

event_date publication_date
72289 2012-07-29       2020-09-25
78320 2016-05-31       2020-09-25
83150 2019-05-05       2020-11-19
80192 2017-07-06       2020-09-25
15189 1986-07-04              NaT
26914 1990-10-28       1992-11-09
2648  1982-09-04       1983-09-04
57936 2004-12-09       2005-02-24
54754 2003-06-03       2003-11-25
38576 1995-10-26       1996-05-30
2482  1982-08-22       1983-08-22
57436 2004-09-05       2005-01-24
78139 2016-04-25       2020-09-25
26478 1990-08-27       1992-09-28
769   1982-04-05       1983-04-05
77892 2016-02-21       2020-09-25
38956 1996-01-14       1996-05-09
24323 1989-10-22       1992-07-29
73791 2013-06-28       2020-09-25
87855 2022-05-27       2022-06-28

In [77]:
#It might be useful to extract the years and the months of the event_date and publication_date which can be useful when doing visualisations
aviation_data_df['event_year'] = aviation_data_df['event_date'].dt.year
aviation_data_df['event_month'] = aviation_data_df['event_date'].dt.month_name()
aviation_data_df['publication_year'] = aviation_data_df['publication_date'].dt.year
aviation_data_df['publication_month'] = aviation_data_df['publication_date'].dt.month_name()

In [78]:
aviation_data_df.head()

event_id investigation_type accident_number event_date  \
0  20001218X45444           Accident      SEA87LA080 1948-10-24   
1  20001218X45447           Accident      LAX94LA336 1962-07-19   
2  20061025X01555           Accident      NYC07LA005 1974-08-30   
3  20001218X45448           Accident      LAX96LA321 1977-06-19   
4  20041105X01764           Accident      CHI79FA064 1979-08-02   

          location        country   latitude   longitude airport_code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built  number_of_engines  \
0              NC6404   Stinson     108-3            No                  1   
1              N5069P     Piper  PA24-180            No                  1   
2              N5142R    Cessna      172M            No                  1   
3              N1168J  Rockwell       112            No                  1   
4               N15NY    Cessna       501            No               <NA>   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

   total_fatal_injuries  total_serious_injuries  total_minor_injuries  \
0                     2                       0                     0   
1                     4                       0                     0   
2                     3                    <NA>                  <NA>   
3                     2                       0                     0   
4                     1                       2                  <NA>   

   total_uninjured weather_condition broad_phase_of_flight   report_status  \
0                0               UNK                Cruise  Probable Cause   
1                0               UNK               Unknown  Probable Cause   
2             <NA>               IMC                Cruise  Probable Cause   
3                0               IMC                Cruise  Probable Cause   
4                0               VMC              Approach  Probable Cause   

  publication_date  event_year event_month  publication_year publication_month  
0              NaT        1948     October               NaN               NaN  
1       1996-09-19        1962        July            1996.0         September  
2       2007-02-26        1974      August            2007.0          February  
3       2000-09-12        1977        June            2000.0         September  
4       1980-04-16        1979      August            1980.0             April

In [79]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 35 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   event_id                88889 non-null  object        
 1   investigation_type      88889 non-null  object        
 2   accident_number         88889 non-null  object        
 3   event_date              88889 non-null  datetime64[ns]
 4   location                88837 non-null  object        
 5   country                 88663 non-null  object        
 6   latitude                34382 non-null  object        
 7   longitude               34373 non-null  object        
 8   airport_code            50249 non-null  object        
 9   airport_name            52790 non-null  object        
 10  injury_severity         87889 non-null  object        
 11  aircraft_damage         85695 non-null  object        
 12  aircraft_category       32287 non-null  object

In [80]:
#due to the null values present in the publication year we need to convert the publication year to interger

aviation_data_df['publication_year'] = aviation_data_df['publication_year'].astype('Int64')

In [81]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 35 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   event_id                88889 non-null  object        
 1   investigation_type      88889 non-null  object        
 2   accident_number         88889 non-null  object        
 3   event_date              88889 non-null  datetime64[ns]
 4   location                88837 non-null  object        
 5   country                 88663 non-null  object        
 6   latitude                34382 non-null  object        
 7   longitude               34373 non-null  object        
 8   airport_code            50249 non-null  object        
 9   airport_name            52790 non-null  object        
 10  injury_severity         87889 non-null  object        
 11  aircraft_damage         85695 non-null  object        
 12  aircraft_category       32287 non-null  object

In [82]:
#Let us reorder our new created columns to be closer to their referenced column
column_names=aviation_data_df.columns.to_list()
column_names

['event_id',
 'investigation_type',
 'accident_number',
 'event_date',
 'location',
 'country',
 'latitude',
 'longitude',
 'airport_code',
 'airport_name',
 'injury_severity',
 'aircraft_damage',
 'aircraft_category',
 'registration_number',
 'make',
 'model',
 'amateur_built',
 'number_of_engines',
 'engine_type',
 'far_description',
 'schedule',
 'purpose_of_flight',
 'air_carrier',
 'total_fatal_injuries',
 'total_serious_injuries',
 'total_minor_injuries',
 'total_uninjured',
 'weather_condition',
 'broad_phase_of_flight',
 'report_status',
 'publication_date',
 'event_year',
 'event_month',
 'publication_year',
 'publication_month']

In [83]:
reorderd_columns=['event_id',
 'investigation_type',
 'accident_number',
 'event_date',
 'event_year',
 'event_month',
 'location',
 'country',
 'latitude',
 'longitude',
 'airport_code',
 'airport_name',
 'injury_severity',
 'aircraft_damage',
 'aircraft_category',
 'registration_number',
 'make',
 'model',
 'amateur_built',
 'number_of_engines',
 'engine_type',
 'far_description',
 'schedule',
 'purpose_of_flight',
 'air_carrier',
 'total_fatal_injuries',
 'total_serious_injuries',
 'total_minor_injuries',
 'total_uninjured',
 'weather_condition',
 'broad_phase_of_flight',
 'report_status',
 'publication_date',
 'publication_year',
 'publication_month']

In [84]:
aviation_data_df=aviation_data_df[reorderd_columns]

In [85]:
aviation_data_df.head()

event_id investigation_type accident_number event_date  event_year  \
0  20001218X45444           Accident      SEA87LA080 1948-10-24        1948   
1  20001218X45447           Accident      LAX94LA336 1962-07-19        1962   
2  20061025X01555           Accident      NYC07LA005 1974-08-30        1974   
3  20001218X45448           Accident      LAX96LA321 1977-06-19        1977   
4  20041105X01764           Accident      CHI79FA064 1979-08-02        1979   

  event_month         location        country   latitude   longitude  \
0     October  MOOSE CREEK, ID  United States        NaN         NaN   
1        July   BRIDGEPORT, CA  United States        NaN         NaN   
2      August    Saltville, VA  United States  36.922223  -81.878056   
3        June       EUREKA, CA  United States        NaN         NaN   
4      August       Canton, OH  United States        NaN         NaN   

  airport_code airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN          NaN        Fatal(2)       Destroyed               NaN   
1          NaN          NaN        Fatal(4)       Destroyed               NaN   
2          NaN          NaN        Fatal(3)       Destroyed               NaN   
3          NaN          NaN        Fatal(2)       Destroyed               NaN   
4          NaN          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built  number_of_engines  \
0              NC6404   Stinson     108-3            No                  1   
1              N5069P     Piper  PA24-180            No                  1   
2              N5142R    Cessna      172M            No                  1   
3              N1168J  Rockwell       112            No                  1   
4               N15NY    Cessna       501            No               <NA>   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

   total_fatal_injuries  total_serious_injuries  total_minor_injuries  \
0                     2                       0                     0   
1                     4                       0                     0   
2                     3                    <NA>                  <NA>   
3                     2                       0                     0   
4                     1                       2                  <NA>   

   total_uninjured weather_condition broad_phase_of_flight   report_status  \
0                0               UNK                Cruise  Probable Cause   
1                0               UNK               Unknown  Probable Cause   
2             <NA>               IMC                Cruise  Probable Cause   
3                0               IMC                Cruise  Probable Cause   
4                0               VMC              Approach  Probable Cause   

  publication_date  publication_year publication_month  
0              NaT              <NA>               NaN  
1       1996-09-19              1996         September  
2       2007-02-26              2007          February  
3       2000-09-12              2000         September  
4       1980-04-16              1980             April

#### Location,Country,Latitudes and longitude

In [86]:
#Let us check the sum of null values in the latitude and longitude
aviation_data_df.isnull().sum()

event_id                      0
investigation_type            0
accident_number               0
event_date                    0
event_year                    0
event_month                   0
location                     52
country                     226
latitude                  54507
longitude                 54516
airport_code              38640
airport_name              36099
injury_severity            1000
aircraft_damage            3194
aircraft_category         56602
registration_number        1317
make                         63
model                        92
amateur_built               102
number_of_engines          6084
engine_type                7077
far_description           56866
schedule                  76307
purpose_of_flight          6192
air_carrier               72241
total_fatal_injuries      11401
total_serious_injuries    12510
total_minor_injuries      11933
total_uninjured            5912
weather_condition          4492
broad_phase_of_flight     27165
report_s

In [87]:
aviation_data_df[['location','country','latitude','longitude']].sample(30)

location        country   latitude    longitude
15310          CANTON, OH  United States        NaN          NaN
2650      NR. MCGRATH, AK  United States        NaN          NaN
65373           Leroy, OH  United States    414235N     0081729W
29100          SUMNER, MS  United States        NaN          NaN
1936     KENNETH CITY, FL  United States        NaN          NaN
57349           LEROY, MN  United States    43.8875   -92.843055
25664        BELGRADE, MT  United States        NaN          NaN
20944         O'BRIEN, FL  United States        NaN          NaN
18749       GREENWOOD, LA  United States        NaN          NaN
4084      BAKERSFIELD, CA  United States        NaN          NaN
5355          WINSLOW, AZ  United States        NaN          NaN
12588         SEATTLE, WA  United States        NaN          NaN
52922            HITE, UT  United States  37.442222  -110.569723
69784     New Orleans, LA  United States    030232N     0090142W
60097           Bouse, AZ  United States  33.931389  -113.995833
74732           Belen, NM  United States    034402N     0106494W
4159   TRINITY CENTER, CA  United States        NaN          NaN
8107           WINONA, MN  United States        NaN          NaN
55384           Miami, FL  United States  25.793055   -80.290555
16576         CENTRAL, UT  United States        NaN          NaN
40673     COUDERSPORT, PA  United States        NaN          NaN
47058        SEMINOLE, TX  United States        NaN          NaN
83339         Antioch, CA  United States    038239N     1215020W
88267      Evansville, IN  United States    038227N     0873143W
18640    WILLIAMSPORT, PA  United States        NaN          NaN
77311          Moneta, VA  United States    037614N     0793550W
78747      Cardington, OH  United States    402945N     0082548W
82324  Frankfurt, Germany        Germany        NaN          NaN
16927        PRESCOTT, AZ  United States        NaN          NaN
82525    Manavgat, Turkey         Turkey        NaN          NaN

In [88]:
#From the above analysis majority of the latitude and longitude values are null values
#We can generate the longitudes and latitudes of the different locations using values in the location and country columns

aviation_data_df[aviation_data_df['country'].isna()]

event_id investigation_type accident_number event_date  \
36     20020917X02410           Accident      MIA82FKA05 1982-01-04   
464    20020917X02358           Accident      MIA82DA062 1982-03-02   
465    20020917X02026           Accident      FTW82DA076 1982-03-02   
725    20020917X02377           Accident      MIA82DA091 1982-03-31   
831    20020917X02069           Accident      FTW82DA127 1982-04-13   
...               ...                ...             ...        ...   
52288  20020322X00387           Accident      DCA02MA029 2002-03-22   
54284  20040528X00699           Accident      DCA03WA031 2003-03-12   
56200  20040528X00697           Accident      DCA04WA026 2004-02-10   
58803  20050616X00790           Incident      DCA05WA073 2005-06-10   
62530  20070518X00582           Incident      DCA07WA043 2007-05-13   

       event_year event_month         location country latitude longitude  \
36           1982     January      SAINT CROIX     NaN      NaN       NaN   
464          1982       March         HUMA CAO     NaN      NaN       NaN   
465          1982       March  MUSTANG BLK A11     NaN      NaN       NaN   
725          1982       March             MOCA     NaN      NaN       NaN   
831          1982       April  WEST DELTA 105D     NaN      NaN       NaN   
...           ...         ...              ...     ...      ...       ...   
52288        2002       March     int'l waters     NaN      NaN       NaN   
54284        2003       March     JOHANNESBURG     NaN      NaN       NaN   
56200        2004    February  Sharjah Airport     NaN      NaN       NaN   
58803        2005        June  Canada/US borde     NaN      NaN       NaN   
62530        2007         May   London Control     NaN      NaN       NaN   

      airport_code        airport_name injury_severity aircraft_damage  \
36             STX  ALEXANDER HAMILTON        Fatal(1)       Destroyed   
464            NaN                 NaN        Fatal(1)       Destroyed   
465            NaN                 NaN       Non-Fatal             NaN   
725           PR05             LABADIE       Non-Fatal     Substantial   
831            NaN                 NaN       Non-Fatal       Destroyed   
...            ...                 ...             ...             ...   
52288          NaN                 NaN       Non-Fatal           Minor   
54284          NaN                 NaN     Unavailable     Substantial   
56200          NaN                 NaN       Fatal(46)       Destroyed   
58803          NaN                 NaN        Incident             NaN   
62530          NaN                 NaN        Incident             NaN   

      aircraft_category registration_number               make     model  \
36             Airplane              N5151U             Cessna       206   
464            Airplane              N2741J             Cessna       150   
465          Helicopter              N1080S               Bell        12   
725            Airplane              N309MJ              Piper    PA-32R   
831          Helicopter               N41AL               Bell      206B   
...                 ...                 ...                ...       ...   
52288               NaN              N234NW  Mcdonnell Douglas  DC-10-30   
54284               NaN                 NaN             Boeing  747-300F   
56200               NaN                 NaN             Fokker        50   
58803               NaN                 NaN                NaN       NaN   
62530               NaN                 NaN             Boeing   737-800   

      amateur_built  number_of_engines    engine_type  \
36               No                  1  Reciprocating   
464              No                  1  Reciprocating   
465              No                  2    Turbo Shaft   
725              No                  1  Reciprocating   
831              No                  1    Turbo Shaft   
...             ...                ...            ...   
52288            No               

In [89]:
# The other dataset contains the full abreviations of some of those states
usstate_codes_df

US_State Abbreviation
0          Alabama           AL
1           Alaska           AK
2          Arizona           AZ
3         Arkansas           AR
4       California           CA
..             ...          ...
57  Virgin Islands           VI
58   Washington_DC           DC
59  Gulf of mexico           GM
60  Atlantic ocean           AO
61   Pacific ocean           PO

[62 rows x 2 columns]

In [90]:
state_dict= dict(zip(usstate_codes_df['Abbreviation'],usstate_codes_df['US_State']))
state_dict

{'AL': 'Alabama',
 'AK': 'Alaska',
 'AZ': 'Arizona',
 'AR': 'Arkansas',
 'CA': 'California',
 'CO': 'Colorado',
 'CT': 'Connecticut',
 'DE': 'Delaware',
 'FL': 'Florida',
 'GA': 'Georgia',
 'HI': 'Hawaii',
 'ID': 'Idaho',
 'IL': 'Illinois',
 'IN': 'Indiana',
 'IA': 'Iowa',
 'KS': 'Kansas',
 'KY': 'Kentucky',
 'LA': 'Louisiana',
 'ME': 'Maine',
 'MD': 'Maryland',
 'MA': 'Massachusetts',
 'MI': 'Michigan',
 'MN': 'Minnesota',
 'MS': 'Mississippi',
 'MO': 'Missouri',
 'MT': 'Montana',
 'NE': 'Nebraska',
 'NV': 'Nevada',
 'NH': 'New Hampshire',
 'NJ': 'New Jersey',
 'NM': 'New Mexico',
 'NY': 'New York',
 'NC': 'North Carolina',
 'ND': 'North Dakota',
 'OH': 'Ohio',
 'OK': 'Oklahoma',
 'OR': 'Oregon',
 'PA': 'Pennsylvania',
 'RI': 'Rhode Island',
 'SC': 'South Carolina',
 'SD': 'South Dakota',
 'TN': 'Tennessee',
 'TX': 'Texas',
 'UT': 'Utah',
 'VT': 'Vermont',
 'VA': 'Virginia',
 'WA': 'Washington',
 'WV': 'West Virginia',
 'WI': 'Wisconsin',
 'WY': 'Wyoming',
 'AS': 'American Samoa',
 'G

In [91]:
#Lets replace the abbreviated state names with full names in the location column
aviation_data_df['location'] = aviation_data_df['location'].replace(state_dict,regex=True)


In [92]:
#Lets convert the locations into title cases
aviation_data_df['location'] = aviation_data_df['location'].str.title()
aviation_data_df.sample(20)

event_id investigation_type accident_number event_date  \
12448  20001214X37235           Accident      MKC85LA170 1985-07-27   
71365  20120218X65637           Accident      ERA12LA184 2012-02-17   
65202  20080905X01387           Accident      NYC08LA291 2008-08-20   
84594  20200330X80034           Accident      CEN20WA135 2020-03-15   
59637  20051108X01802           Accident      MIA06CA009 2005-10-02   
8754   20001214X39795           Accident      ANC84LA102 1984-06-24   
70567  20110816X84113           Accident      ERA11CA453 2011-08-13   
9691   20001214X41107           Accident      LAX84FU033 1984-09-07   
9837   20001214X41035           Accident      DEN84FU011 1984-09-20   
61176  20060923X01382           Accident      LAX06CA252 2006-08-03   
5302   20001214X43795           Accident      LAX83FVM09 1983-07-03   
55389  20031003X01653           Accident      FTW03LA204 2003-08-25   
10980  20001214X35887           Accident      DEN85LA089 1985-03-01   
37668  20001207X03937           Accident      FTW95LA268 1995-07-04   
41498  20001208X07678           Accident      CHI97LA098 1997-04-02   
34648  20001206X01002           Accident      SEA94LA078 1994-03-08   
62851  20070711X00909           Accident      CHI07LA191 2007-07-04   
44291  20001211X10674           Accident      NYC98LA140 1998-07-02   
32477  20001211X12221           Accident     NYC93LA074A 1993-04-07   
60283  20060407X00407           Accident      SEA06CA061 2006-03-05   

       event_year event_month                                       location  \
12448        1985        July                                     Leon, Iowa   
71365        2012    February                   Fayetteville, North Carolina   
65202        2008      August                            Baltimore, Maryland   
84594        2020       March                Tepalcatepec, Michoacan, Mexico   
59637        2005     October                       Punta Goregonda, Florida   
8754         1984        June                       Bearkansas Creek, Alaska   
70567        2011      August                         West Liberty, Kentucky   
9691         1984   September               Lowerlouisianalaskae, California   
9837         1984   September                   Brhode Islandghton, Colorado   
61176        2006      August  San Anorth Dakotareamerican Samoa, California   
5302         1983        July                             Oceano, California   
55389        2003      August           Wisconsinndianansboregono, Louisiana   
10980        1985       March                      Rhode Islandchfield, Utah   
37668        1995        July        Farkansasmichiganndianagton, New Mexico   
41498        1997       April                      Louisianafayette, Indiana   
34648        1994       March                               Heber City, Utah   
62851        2007        July                            Johionson, Nebraska   
44291        1998        July         Massachusettshopennsylvaniac, New York   
32477        1993       April              Pennsylvaniawtucket, Rhode Island   
60283        2006       March              Saindianat Pennsylvaniaul, Oregon   

             country   latitude    longitude airport_code  \
12448  United States        NaN          NaN          NaN   
71365  United States    034596N     0785322W          FAY   
65202  United States    039110N     0763936W          BWI   
84594         Mexico    019955N     1025029W          NaN   
59637  United States  26.922778   -81.995277         KPGD   
8754   United States        NaN          NaN          NaN   
70567  United States    375452N     0083158W          9I3   
9691   United States        NaN          NaN          NaN   
9837   United States        NaN          NaN          NaN   
61176  United States  38.146111  -120.648055          CPU   
5302   United States        NaN          NaN          NaN   
55389  United States  32.153056   -91.698333          NaN   
10980  United States        NaN          NaN   

In [93]:
aviation_data_df['location'].isna().sum()

52

In [94]:
# Function to change longitude and latitude to float type
def dms_to_dd(dms_str):
    if pd.isna(dms_str):
        return np.nan
    
    # Regular expression to extract degrees, minutes, and seconds
    dms_pattern = re.compile(r'(\d{2,3})(\d{2})(\d{2})([NSEW])')
    match = dms_pattern.match(dms_str)
    if not match:
        return np.nan
    
    degrees, minutes, seconds, direction = match.groups()
   
    dd = float(degrees) + float(minutes)/60 + float(seconds)/3600
    
    if direction in ['S', 'W']:
        dd *= -1
    
    return dd


In [95]:

aviation_data_df['latitude'] = aviation_data_df['latitude'].apply(lambda x: dms_to_dd(x) if isinstance(x, str) and re.match(r'\d{6}[NSEW]', x) else pd.to_numeric(x, errors='coerce'))
aviation_data_df['longitude'] = aviation_data_df['longitude'].apply(lambda x: dms_to_dd(x) if isinstance(x, str) and re.match(r'\d{7}[NSEW]', x) else pd.to_numeric(x, errors='coerce'))

In [96]:
aviation_data_df[['latitude','longitude']].sample(20)

latitude   longitude
34813        NaN         NaN
8209         NaN         NaN
62854  38.256111 -119.223334
935          NaN         NaN
61898  -3.916667  118.216667
19316        NaN         NaN
1189         NaN         NaN
3126         NaN         NaN
59372  33.171945  -97.823611
33292        NaN         NaN
12410        NaN         NaN
33364        NaN         NaN
77629  33.964167 -112.809444
70858  27.906111   -8.405000
2693         NaN         NaN
21307        NaN         NaN
51062        NaN         NaN
2594         NaN         NaN
26080        NaN         NaN
86044        NaN         NaN

In [97]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 35 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   event_id                88889 non-null  object        
 1   investigation_type      88889 non-null  object        
 2   accident_number         88889 non-null  object        
 3   event_date              88889 non-null  datetime64[ns]
 4   event_year              88889 non-null  int64         
 5   event_month             88889 non-null  object        
 6   location                88837 non-null  object        
 7   country                 88663 non-null  object        
 8   latitude                34382 non-null  float64       
 9   longitude               34373 non-null  float64       
 10  airport_code            50249 non-null  object        
 11  airport_name            52790 non-null  object        
 12  injury_severity         87889 non-null  object